src/classes/frDonation.cls

```diff
@@ -53,84 +55,126 @@ public class frDonation extends frModel implements frSyncable {
                                                 frUtil.Entity.CAMPAIGN, campaignId);
                 }
             }
         }
-        
+        //connect subscription
         String subscriptionId = String.valueOf(request.get('subscriptionId'));
         if (String.isNotBlank(subscriptionId)) {
-            List<Subscription__c> subscriptions = [SELECT Id FROM Subscription__c WHERE fr_ID__c = :subscriptionId];
+            List<SObject> subscriptions = Database.query('SELECT Id FROM ' + subsciptionObjName + ' WHERE fr_ID__c = :subscriptionId');
             if (subscriptions.size() > 0) {
-                o.Subscription__c = subscriptions.get(0).Id;
+                o.put('GiftCommitmentId', subscriptions.get(0).get('Id'));
             } else {
                 frUtil.logRelationshipError(getFrType(), funraiseId, 
                                             frUtil.Entity.SUBSCRIPTION, subscriptionId);
             }
         }
-        Opportunity opp = this.o;
-        
-        
+
+        //connect pledge
         Boolean isPledge = Boolean.valueOf(request.get('pledge'));
-        Boolean alreadyMappedToPledge = [SELECT COUNT() FROM Opportunity 
-                                         WHERE fr_Id__c = :funraiseId AND Funraise_Pledge__c != null] > 0;
-        if(!isPledge && !alreadyMappedToPledge) {
-            Pledge__c pledge = frPledge.findActive(opp.fr_Donor__c, opp.CloseDate);
-            opp.Funraise_Pledge__c = pledge != null ? pledge.Id : null;
+        if(frUtil.hasNPCobjects()) {
+            List<SObject> alreadyMappedToPledgeList= Database.query('SELECT id FROM GiftTransaction WHERE fr_Id__c = :funraiseId AND Funraise_Pledge__c != null');
+            Boolean alreadyMappedToPledge = alreadyMappedToPledgeList.size() > 0;
+            if(!isPledge && !alreadyMappedToPledge) {
+                Pledge__c pledge = frPledge.findActiveAcc((Id) o.get('DonorId'), (Date) o.get('CheckDate'));
+                NPC_Setting__c npcSetting = NPC_Setting__c.getOrgDefaults();
+                if (npcSetting != null && pledge != null) {
+                    String pledgeId = npcSetting.NPCFieldName__c;
+                    o.put('Funraise_GT_Pledge__c', pledge.id);
+                } else {
+                    System.debug('NPC Setting not found at the organization level.');
+                }
+            }
         }
-        
+        else {
+            Opportunity opp = (Opportunity) this.o;
+            Boolean alreadyMappedToPledge = [SELECT COUNT() FROM Opportunity 
+                                            WHERE fr_Id__c = :funraiseId AND Funraise_Pledge__c != null] > 0;
+            if(!isPledge && !alreadyMappedToPledge) {
+                Pledge__c pledge = frPledge.findActive(opp.fr_Donor__c, opp.CloseDate);
+                opp.Funraise_Pledge__c = pledge != null ? pledge.Id : null;
+            }
+        }
+
+        //upsert
         try {
-            Database.upsert(opp, Opportunity.Fields.fr_Id__c, true);
```

Comment:
Removing this upsert on the Funraise ID means we will create duplicate opportunities when the same Funraise record is synced multiple times, because it doesn't look like there is a query to find an existing opp for the same fr_Id__c value,

Reply:
Added this back for upserting opportunities! we check for existing gift transactions but this can be changed if needed!